fix splitting sharp edges with active cell scalars

akaszynski opened the pull request:
Splitting sharp edges fails for polydata when the active scalars are cells:

```py
import pyvista as pv

cube = pv.Cube()
cube.cell_data['cell_id'] = range(cube.n_cells)
cube = cube.triangulate().subdivide(2)

cube.plot(smooth_shading=True, split_sharp_edges=True)
```

```
File ~/python/pyvista/pyvista/core/datasetattributes.py:769, in DataSetAttributes._prepare_array(self, data, name, deep_copy)
    767     data = tmparray
    768 if data.shape[0] != array_len:
--> 769     raise ValueError(f'data length of ({data.shape[0]}) != required length ({array_len})')
    771 if data.dtype == np.bool_:
    772     self.dataset.association_bitarray_names[self.association.name].add(name)

ValueError: data length of (192) != required length (98)
```

The fix is to simply not track cell indices since they do not change when splitting sharp edges.

akaszynski commented:
This broke our docs at [PyMAPDL documentation](https://mapdldocs.pyansys.com/examples/index.html) and I'd like to have this out for the next release (end of June?). Appreciate a review when anyone gets a chance.

adeak commented on pyvista/plotting/_plotting.py:
```diff
@@ -78,14 +78,13 @@ def prepare_smooth_shading(mesh, scalars, texture, split_sharp_edges, feature_an
-                # we must track the original IDs with our own array
+            if has_scalars and use_points:
+                # we must track the original IDs with our own array since point
+                # indices may have changed
                 indices_array = '__orig_ids__'
-                if use_points:
-                    arr_sz = mesh.n_points
-                else:
-                    arr_sz = mesh.n_cells
+                arr_sz = mesh.n_points
                 mesh.point_data[indices_array] = np.arange(arr_sz, dtype=np.int32)
+
```
If I understand correctly, the scheme is that
1. in case of `UnstructuredGrid` input we don't have to do this, because we already have the `'vtkOriginalPointIds'` array that we'll use later to restore the split scalars
2. in case of `PolyData` input with active cells we don't have to do this, because edge splitting is actually point splitting, so cells (and their scalars) are unchanged
3. so we only need to keep track of the original point indices in case the input was already a `PolyData` (hence no `'vtkOriginalPointIds'`) and it has active point scalars (hence merely splitting would mess with the scalars).

This raises the question, though: couldn't/shouldn't we do all this inside `compute_normals`, which is the method that actually changes the size of the mesh in case of a split? We'd have to handle every point array in the same way, but that would kind of make sense to me.

Anyway, if I understand the above correctly, this seems good to me.

akaszynski replied:
> 1. in case of `UnstructuredGrid` input we don't have to do this, because we already have the `'vtkOriginalPointIds'` array that we'll use later to restore the split scalars

Yes. More specifically, `extract_surface` on any dataset will return a `PolyData` containing the `'vtkOriginalPointIds'` array.



> 2. * in case of `PolyData` input with active cells we don't have to do this, because edge splitting is actually point splitting, so cells (and their scalars) are unchanged
> 3. * so we only need to keep track of the original point indices in case the input was already a `PolyData` (hence no `'vtkOriginalPointIds'`) and it has active point scalars (hence merely splitting would mess with the scalars).

Correct on both accounts.

> This raises the question, though: couldn't/shouldn't we do all this inside compute_normals, which is the method that actually changes the size of the mesh in case of a split?

That's not a bad proposal. In fact, there are cases in which I've done this in the past on personal projects. Implemented in:
915230a799755f8138cc2fbb04811d6b2877c1ef

> We'd have to handle every point array in the same way, but that would kind of make sense to me.

I'd argue that we only have to this for our new ID tracking array as this is already being done internally by VTK.

adeak replied:
> I'd argue that we only have to this for our new ID tracking array as this is already being done internally by VTK.

So the remaining question is why we have to do this whole tracking dance in the first place. And I suspect the answer is because the scalars might be passed as an explicit array. Otherwise (i.e. when the scalars come from point data) VTK would be handling it correctly anyway. But I can see how making use of that would not be feasible with the current implementation.

akaszynski replied:
> I suspect the answer is because the scalars might be passed as an explicit array.

Correct. In other projects using `pyvista`, I've had to generate the ID array separately, so I think any work here will also be used outside `_plotter.py`.

adeak replied:
OK, thanks for clearing that up. I'll review the updated state tomorrow.

akaszynski commented on pyvista/core/filters/poly_data.py:
```diff
@@ -1311,7 +1311,9 @@ def compute_normals(
         split_vertices : bool, optional
-            Splitting of sharp edges. Defaults to ``False``.
+            Splitting of sharp edges. Defaults to ``False``. Indices to the
+            original points is tracked in the ``"pyvistaNormOriginalPointIds"
```
```suggestion
            original points is tracked in the ``"pyvistaNormOriginalPointIds"``
```

adeak reviewed (approved):
I've added a few small changes, please check. Otherwise this PR looks good to me. That being said, two points:

1. in `prepare_smooth_shading` we talk about Phong shading, but in `compute_normals` we talk about Gouraud. Shouldn't these refer to the same method? I know there was a recent-ish switch from one to the other... but perhaps that was only for PBR?
2. right now we use `np.int32` for the original point IDs, presumably to spare memory But `np.iinfo(np.int32).max` says this goes up to `2147483647`, i.e. 2 billion points. If I estimated correctly, even with double-precision coordinates such a mesh (sans data) would amount to 48 GB of memory, which is arguably not that far-fetched these days. How about we make use of the fact that point indices are non-negative, and use `np.uint32` instead? Or `np.uint64` but I can see how double the memory for this auxiliary array can become a problem... Also cc @MatthewFlamm who usually has strong points about memory.

akaszynski commented:
> 1. in `prepare_smooth_shading` we talk about Phong shading, but in `compute_normals` we talk about Gouraud. Shouldn't these refer to the same method? I know there was a recent-ish switch from one to the other... but perhaps that was only for PBR?

Good catch. We're using `prop.SetInterpolationToPhong()` and switched from Gouraud shading a while back. In fact, there's inconsistent documentation within `plotter.py` that I removed in 962537a914f05811044f8caf2e67f3f9c0379996.

> So I think this is fine, but if unsigned integers works why not?

That's a reasonable point, and as much as I'd like to improve the memory efficiency of pyvista, I don't want to stay too far from the way that VTK does things.

Speaking of that, I picked `np.int32` because of memory efficiency, but as it turns out, VTK uses int32 for Windows and int64 for Linux. This is the int type used for `vtk.vtkIdTypeArray` and we obtain it from `pyvista.utilities.misc._get_vtk_id_type`. For the sake of consistency, I'm going to use that type, even if it's unnecessarily large. As @MatthewFlamm, it's an opt-in.

Implemented that change in 8896330b01a46ee424b87a47f7611a7d0e557d36.

adeak commented:
> That's a reasonable point, and as much as I'd like to improve the memory efficiency of pyvista, I don't want to stay too far from the way that VTK does things.

That's fair.

> Speaking of that, I picked np.int32 because of memory efficiency, but as it turns out, VTK uses `int32` for Windows and `int64` for Linux.

For what it's worth this sounds a lot like just "numpy default int". Int the size of the word size (`int32` on 32-bit systems and `int64` on 64-bit systems) _except_ 64-bit Windows which also has `int32`. And the reason is [dark history involving the size of `long`](https://github.com/numpy/numpy/issues/9464).

adeak commented on pyvista/plotting/plotting.py:
```diff
@@ -1843,10 +1843,10 @@ def add_mesh(
-            Gouraud or Phong shading algorithm.  When ``False``, use
-            flat shading.  Automatically enabled when ``pbr=True``.
-            See :ref:`shading_example`.
+            Enable smooth shading when ``True`` using either the the Phong
```
```suggestion
            Enable smooth shading when ``True`` using the Phong
```

akaszynski commented:
Note: I'm skipping a test for MacOS. Now that we're running image cache verification, I expect to run into these issues more often, especially on MacOS.
